### PR TITLE
AMLII-1110 - Fix Additional UseSSL setting for additional endpionts in logs agent

### PR DESCRIPTION
**Pull request by soberpeach**
### What does this PR do?

This PR fixes the `logs_no_ssl` setting for additional logs endpoints (which controls the `UseSSL` setting of the endpoint) to respect the value configured in `datadog.yaml` instead of inheriting from the main endpoint. It also adds a warning for a future upcoming breaking change to the way `dd_url` and `logs_no_ssl` interact.

### Motivation

This has been broken since inception.

### Possible Drawbacks / Trade-offs

None that I can think of. If customers relied on this behavior before they can still achieve the same thing, but now they have the option to utilize the `logs_no_ssl` setting for additional endpoints.

### Describe how to test/QA your changes

Run the agent with logs enabled and add additional endpoints in the `logs_configuration` like so:

```
logs_config:
  use_http: true
  additional_endpoints:
  - api_key: "apiKey2"
    Host: "randval"
    Port: 443
    is_reliable: true
    logs_no_ssl: false
```

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by paulcacheux — APPROVED** [review]

File: comp/logs/agent/config/config.go [hunk 1]
```diff
@@ -17,6 +17,7 @@ import (
 	pkgConfig "github.com/DataDog/datadog-agent/pkg/config"
 	"github.com/DataDog/datadog-agent/pkg/config/utils"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
+	pointer "github.com/DataDog/datadog-agent/pkg/util/pointer"
```
Comment (rayz) [inline comp/logs/agent/config/config.go:20]: is it necessary to have this alias since the filename ends with `pointer` already?

File: releasenotes/notes/fixed-ssl-settings-for-addition-endpoints-4e63b4271b21ff17.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+fixes:
+  - |
+    Fixed logs agent additional endpoint to respect their
+    logs_no_ssl setting.
```
Comment (brett0000FF) [inline releasenotes/notes/fixed-ssl-settings-for-addition-endpoints-4e63b4271b21ff17.yaml:12]:
```suggestion
    Fixed Logs Agent additional endpoints to respect their
    logs_no_ssl setting.
```

**Review by brett0000FF — APPROVED** [review]

File: comp/logs/agent/config/endpoints.go [hunk 3]
```diff
@@ -37,9 +37,9 @@ type Endpoint struct {
 	APIKey                  string `mapstructure:"api_key" json:"api_key"`
 	Host                    string
 	Port                    int
-	UseSSL                  bool
-	UseCompression          bool `mapstructure:"use_compression" json:"use_compression"`
-	CompressionLevel        int  `mapstructure:"compression_level" json:"compression_level"`
+	UseSSL                  *bool `mapstructure:"logs_no_ssl" json:"logs_no_ssl"`
```
Comment (gh123man) [inline comp/logs/agent/config/endpoints.go:40]:
```suggestion
	UseSSL                  *bool `mapstructure:"use_ssl" json:"use_ssl"`
```
I think this should be `use_ssl`. `logs_no_ssl` is the inverse of the underlying setting. 

I it may be safe to rename since it never worked to begin with (as a public config), but I think we should probably leave it as is.

File: comp/logs/agent/config/endpoints_test.go [hunk 4]
```diff
@@ -488,6 +488,37 @@ func (suite *EndpointsTestSuite) TestIsReliableDefaultTrue() {
...
 
+func (suite *EndpointsTestSuite) TestUseSSLDefaultTrue() {
+	var (
+		endpoints *Endpoints
+		err       error
+	)
+
+	suite.config.SetWithoutSource("logs_config.additional_endpoints", []map[string]interface{}{
+		{
+			"host":    "a",
```
Comment (gh123man) [inline comp/logs/agent/config/endpoints_test.go:499]: can we also add a case (or another test) that handles:
`logs_config.logs_no_ssl: true`
`logs_config.logs_no_ssl: false`
`logs_config.logs_no_ssl:` unset

with each permutation of `use_ssl` at the endpoint level? This way we can verify that the setting is correctly overridden (set to the same as the main endpoint) when unset, and overridden by the more specific endpoints settings when needed.

File: releasenotes/notes/fixed-ssl-settings-for-addition-endpoints-4e63b4271b21ff17.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+#
+# Each section note must be formatted as reStructuredText.
+---
+fixes:
+  - |
+    Fixed Logs Agent additional endpoints to respect their
+    logs_no_ssl setting.
+other:
+  - |
+    Added warning for upcoming breaking change to logs_no_ssl
```
Comment (gh123man) [inline releasenotes/notes/fixed-ssl-settings-for-addition-endpoints-4e63b4271b21ff17.yaml:15]: I'd maybe reword this to something like:

`Added a warning when ``logs_no_ssl`` is set and the url contains an http or https prefix. ``logs_no_ssl`` will take precedence over the prefix in a future agent version.`

File: comp/logs/agent/config/config.go [hunk 6]
```diff
@@ -281,6 +292,9 @@ type defaultParseAddressFunc func(string) (host string, port int, err error)
 
 func parseAddressWithScheme(address string, defaultNoSSL bool, defaultParser defaultParseAddressFunc) (host string, port int, useSSL bool, err error) {
 	if strings.HasPrefix(address, "https://") || strings.HasPrefix(address, "http://") {
+		if strings.HasPrefix(address, "https://") && !defaultNoSSL {
+			log.Warn("dd_url set to a URL with an HTTPS prefix and logs_no_ssl set to true. These are conflicting options and next release will introduce a breaking change where logs_no_ssl overrides dd_url.")
```
Comment (gh123man) [inline comp/logs/agent/config/config.go:296]:
```suggestion
			log.Warn("dd_url is set to a URL with an HTTPS prefix and logs_no_ssl is set to true. These are conflicting options. In a future release logs_no_ssl will override the dd_url prefix")
```

File: comp/logs/agent/config/config.go [hunk 7]
```diff
@@ -243,12 +250,16 @@ func BuildHTTPEndpointsWithConfig(coreConfig pkgConfig.Reader, logsConfig *LogsC
...
+		*main.UseSSL = useSSL
 	}
 
 	additionals := logsConfig.getAdditionalEndpoints()
 	for i := 0; i < len(additionals); i++ {
-		additionals[i].UseSSL = main.UseSSL
+		if additionals[i].UseSSL == nil {
+			additionals[i].UseSSL = main.UseSSL
+		} else {
+			*additionals[i].UseSSL = !*additionals[i].UseSSL
```
Comment (gh123man) [inline comp/logs/agent/config/config.go:261]: I don't think we want to flip this - I think this can be removed when the additional endpoint setting is renamed (from `logs_no_ssl` -> `use_ssl`)?

File: comp/logs/agent/config/config.go [hunk 8]
```diff
@@ -170,12 +172,16 @@ func buildTCPEndpoints(coreConfig pkgConfig.Reader, logsConfig *LogsConfigKeys)
...
+		*main.UseSSL = !logsConfig.devModeNoSSL()
 	}
 
 	additionals := logsConfig.getAdditionalEndpoints()
 	for i := 0; i < len(additionals); i++ {
-		additionals[i].UseSSL = main.UseSSL
+		if additionals[i].UseSSL == nil {
+			additionals[i].UseSSL = main.UseSSL
+		} else {
+			*additionals[i].UseSSL = !*additionals[i].UseSSL
```
Comment (gh123man) [inline comp/logs/agent/config/config.go:183]: Same here

**Review by gh123man — CHANGES_REQUESTED** [review]
I think this is close! Just a few comments to tie things up.

File: comp/logs/agent/config/config.go [hunk 9]
```diff
@@ -217,6 +221,7 @@ func BuildHTTPEndpointsWithConfig(coreConfig pkgConfig.Reader, logsConfig *LogsC
 	} else {
 		main.Version = EPIntakeVersion1
 	}
+	main.UseSSL = pointer.Ptr(defaultNoSSL)
```
Comment (gh123man) [inline comp/logs/agent/config/config.go:224]: NIT: maybe we can move this into the `main := Endpoint{` block similar to `buildTCPEndpoints`? (also can do `&defaultNoSSL` and omit the pointer util)

File: comp/logs/agent/config/config_test.go [hunk 10]
```diff
@@ -575,6 +575,8 @@ func getTestEndpoint(host string, port int, ssl bool) Endpoint {
 		Protocol:         "test-proto",
 		Origin:           "test-source",
 	}
+	e.UseSSL = pointer.Ptr(ssl)
```
Comment (gh123man) [inline comp/logs/agent/config/config_test.go:578]: NIT: I think you can put this back in the `Endpoint{` block.

File: pkg/security/utils/endpoint.go [hunk 11]
```diff
@@ -16,7 +16,7 @@ import (
 func GetEndpointURL(endpoint logsconfig.Endpoint, uri string) string {
 	port := endpoint.Port
 	var protocol string
-	if endpoint.UseSSL {
+	if endpoint.UseSSL == nil || *endpoint.UseSSL {
```
Comment (gh123man) [inline pkg/security/utils/endpoint.go:19]: NIT: can we use `GetUseSSL()` here?

**Review by gh123man — APPROVED** [review]
LGTM - just some nits :) 
Nice work 🚀
